File: 0003-LongestSubstring/LongestSubstring.java

import java.util.HashMap;

public class LongestSubstring {
    class Solution {
        public int lengthOfLongestSubstring(String s) {
            int maxLength = 0, left = 0, right = 0, windowLength = 0;
            HashMap<Character, Integer> map =  new HashMap<>();
            while(right < s.length()){
                if(map.get(s.charAt(right)) != null && map.get(s.charAt(right)) >= left){
                    left = map.get(s.charAt(right)) + 1;
                }
                windowLength = right - left + 1;
                maxLength = maxLength > windowLength ? maxLength : windowLength;
                map.put(s.charAt(right), right);
                right++;
            }
            return maxLength;
        }
    }
}

File: 0003-LongestSubstring/LongestSubstring.cpp
#include <string>
using namespace std;

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int maxLength = 0, windowLength, left = 0;
        int seen[128];
        fill(seen, seen + 128, -1);
        for(int right = 0; right < s.length(); right++){
            char c = s.at(right);
            if(seen[c] != -1 && seen[c] >= left){
                left = seen[c] + 1;
            }
            windowLength = right - left + 1;
            maxLength = maxLength > windowLength ? maxLength : windowLength;
            seen[c] = right;
        }
        return maxLength;
    }
};